Hosting onboarding: add site options step to `/setup/new-hosted-site`

zaguiini opened the pull request:
Closes https://github.com/Automattic/dotcom-forge/issues/2360.

## Proposed Changes

This PR adds the data center picker, along with a site title suggestion, to the new `/hosting` onboarding flow. See the issue for details.

![image](https://github.com/Automattic/wp-calypso/assets/26530524/47b7f884-71a1-4103-a7d2-20d3dc13167f)

## Testing Instructions

1. Patch your sandbox with this diff: D110494-code;
2. Checkout this branch;
3. Browse `/setup/new-hosted-site`;
4. Verify that the site title gets auto-populated with a suggestion;
5. Verify that the recommended data picker is selected, but you can pick a custom one;
6. Buy the plan;
7. Verify that the site title is the same as the one suggested in the first step and that the domain slug is related to the suggestion;
8. Verify that the picked data center was communicated correctly*.

*You can use this snippet to log the picked data center right after checkout. Add this to `0-sandbox` **before pressing the Pay button**:

```php
function sandbox_activate_product( $blog_id, $user_id, $product_id, $txn_id, $verb, $meta, $volume, $extra ) {
	l( $extra );
}

add_action( 'activate_product', 'sandbox_activate_product', 9, 8 );
```

And you should see `{ extra: { geo_affinity: 'chosen_data_center' } }` on your sandbox logs.

zaguiini committed:
Add data center picker to site options step

zaguiini committed:
Add site options image

zaguiini committed:
Fetch suggestions

zaguiini committed:
Add suggestion to the step if site title is empty

zaguiini committed:
Wire site title preference

zaguiini committed:
Only reset store if browsing the first step

zaguiini committed:
Wire data center preference

javierarce commented:
Could we hide the dropdown when the optimal data center is selected? We don't need to show it by default and it would make this screen look nicer.

![image](https://github.com/Automattic/wp-calypso/assets/4933/b2434a57-f1f8-44e1-8f7e-57dcdb5b3726)

zaguiini committed:
Hide data center picker if choosing the recommended one

zaguiini committed:
Move the data center picker to its own block

mpkelly reviewed (approved):
It works as advertised. Code looks fine although the conditional logic is a bit complex in places. 

I agree that removing illustrations feels like a separate PR unrelated to this one.

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/plans/index.tsx:
```diff
@@ -27,13 +35,13 @@ const plans: Step = function Plans( { navigation, flow } ) {
 
 	const shouldHideBackButton = () => {
-		if ( flow === DOMAIN_UPSELL_FLOW ) {
+		if ( flow === DOMAIN_UPSELL_FLOW || isHostingSiteCreationFlow( flow ) ) {
 			return false;
 		}
 		return true;
```
Would it be clearer expressed like this?

```suggestion
		return flow !== DOMAIN_UPSELL_FLOW && ! isHostingSiteCreationFlow( flow );
```

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/plans/index.tsx:
```diff
@@ -9,8 +15,10 @@ const plans: Step = function Plans( { navigation, flow } ) {
-	const handleSubmit = () => {
-		const providedDependencies: ProvidedDependencies = {};
+	const handleSubmit = ( plan: MinimalRequestCartProduct | null ) => {
+		const providedDependencies: ProvidedDependencies = {
+			...plan,
+		};
```
Do the plan properties really want to be spread into the dependencies object? I know it's unlikely, but imagine if someone added a `goToCheckout` prop to the plan object. We'd end up overriding it with our own `goToCheckout flag.

It feels like the values submitted by a step should be distinct from the plans's own properties. i.e.

```suggestion
		const providedDependencies: ProvidedDependencies = { plan };
```

p-jackson commented on client/landing/stepper/declarative-flow/internals/steps-repository/site-options/style.scss:
```diff
@@ -149,3 +149,50 @@
+		}
+	}
+
+	.data-center-picker {
+		color: var(--color-text);
+	}
+
+	.data-center-picker__description {
+		white-space: nowrap;
+
+		&,
+		& a {
+			font-size: $font-body-extra-small;
+			color: var(--color-text);
+		}
+
+		a {
+			text-decoration: underline;
+		}
+	}
+
+	.data-center-picker__options {
+		font-size: $font-body-small;
+
+		.form-select {
+			width: fit-content;
+			border-radius: 4px;
+			font-size: $font-body-small;
+		}
+	}
```
It feels like some of these styles belong in the `client/blocks/data-center-picker` code, rather than in the step repository. I understand that some of the theming we want to do in the stepper framework, but wouldn't some of the rules (e.g. `.form-select` and the wrapping rule) go in the component itself?

In fact there seems to be something wrong with the wrapping rule. It's not allowing the description to fit correctly on small screens.

![wrap](https://github.com/Automattic/wp-calypso/assets/1500769/79d3a985-8820-47e1-b111-7004f2a64139)

p-jackson commented on client/landing/stepper/declarative-flow/hosting-site-creation-flow.ts:
```diff
@@ -60,6 +84,20 @@ const hosting: Flow = {
+
+		useEffect(
+			() => {
+				if ( currentStepSlug === 'options' ) {
+					resetOnboardStore();
+				}
```
This does something when the user navigates directly to `/setup/new-hosted-site/options`, but not when the user goes to `/setup/new-hosted-site` (which is the usual URL we use to start a flow). That's because `currentStepSlug === undefined`.

In my experience, deciding when is the best time to reset the flow is one of the trickiest parts of the onboarding process. No easy answers 🤷 Reseting the flow for each fresh load of the `options` step seems a good place to start though.

The `free` flow does it [for every page load](https://github.com/Automattic/wp-calypso/blob/51fb46a0166b32b40ef986bc5d898b13775c3233/client/landing/stepper/declarative-flow/free.ts#L38) but that seems too aggressive for me. If the user refreshes on the plans step we should retain their geo affinity etc.

p-jackson reviewed (commented):
### Plans back button not working

![back](https://github.com/Automattic/wp-calypso/assets/1500769/739aa494-2f9e-4406-896f-13c5b7a8f375)

This button doesn't work. That's because `plans` step's `handleGoBack()` function is special cased to only do something for the `DOMAIN_UPSELL_FLOW`. @alshakero now that all navigation is modelled as "submitting" a step, should there be some generic understanding for what it means to go back? e.g. `submit?.( { back: true } );`
Without it, every flow is going to have to customise the plan step with it's own bespoke way of going backwards. I don't _think_ it'll break anything if we introduce it now, since we won't be introducing a new code path for flows where the back button is hidden, and basically all flows hide the back button.

### Using the site options step

Just a thought that the site options step seems to be getting pretty darn big. There's so many different modes and permutations, and it makes me wonder whether it's cleaner and easier to test if we just copy-paste the existing site options step i.e. have a new step that only includes the two options we care about.
The down side is we'd need to come up with our own slug.

p-jackson reviewed (commented):
### Plans back button not working

![back](https://github.com/Automattic/wp-calypso/assets/1500769/739aa494-2f9e-4406-896f-13c5b7a8f375)

This button doesn't work. That's because `plans` step's `handleGoBack()` function is special cased to only do something for the `DOMAIN_UPSELL_FLOW`. @alshakero now that all navigation is modelled as "submitting" a step, should there be some generic understanding for what it means to go back? e.g. `submit?.( { back: true } );`
Without it, every flow is going to have to customise the plan step with it's own bespoke way of going backwards. I don't _think_ it'll break anything if we introduce it now, since we won't be introducing a new code path for flows where the back button is hidden, and basically all flows hide the back button.

### Using the site options step

Just a thought that the site options step seems to be getting pretty darn big. There's so many different modes and permutations, and it makes me wonder whether it's cleaner and easier to test if we just copy-paste the existing site options step i.e. have a new step that only includes the two options we care about.
The down side is we'd need to come up with our own slug.

alshakero commented:
Thanks for the ping!

> This button doesn't work. That's because plans step's handleGoBack() function is special cased to only do something for the DOMAIN_UPSELL_FLOW. @alshakero now that all navigation is modelled as "submitting" a step, should there be some generic understanding for what it means to go back? e.g. `submit?.( { back: true } )`;

Yes, that sounds like a good idea. Although I feel it's like I'd be tricking myself around not using `goBack`. Maybe I'm mistaken and `goBack` is a legitimate need, and we shouldn't depreciate it? `goBack` isn't too bad because it doesn't make the step less reusable and doesn't leak information about the flow into the step.

We can just remove [this](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/types.ts#L20-L22) if you agree. WDYT?

zaguiini committed:
Reset flow for every fresh load

zaguiini committed:
Submit plan object instead of spreading the properites

zaguiini committed:
Fix go back behavior

zaguiini committed:
Branch site options to multiple components according to the chosen flow

zaguiini committed:
Wrap data center description on smaller screens

zaguiini committed:
Separate data picker styles

zaguiini commented:
Thanks for all the feedback, @p-jackson! Please let's give this PR another round of review.

>We can just remove [this](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/types.ts#L20-L22) if you agree. WDYT?

Sounds good @alshakero. To not mix concerns, it'd make more sense to do this on a separate PR.

p-jackson commented:
> Maybe I'm mistaken and `goBack` is a legitimate need, and we shouldn't depreciate it? `goBack` isn't too bad because it doesn't make the step less reusable and doesn't leak information about the flow into the step.

@alshakero Fair enough, insisting on `submit( { back: true } )` over `goBack()` is probably just forcing people to jump through hoops when they're obviously the same thing 😄 

FWIW, the thing I like about only having `submit()` is that it forces everyone to think of their `useStepNavigation` hook as a finite-state machine where going back isn't anything special. So it was a good proposal.

javierarce commented:
Nice work! I noticed a few areas for improvement:

- We should remove the vertical scrolling on the page.
- In other onboarding flows, we use a different pattern for required steps. For instance, [in the newsletter setup](https://wordpress.com/setup/newsletter/newsletterSetup), the submit button remains active, but we display an error message if the user fails to provide a newsletter name. We also follow this pattern in the importing stage of other onboardings. To maintain consistency, let's enable the continue button and show an error message if the user attempts to proceed without a title.
- It could be related to using Calypso live for testing, but the continue button's background color is pink instead of blue.
- The labels' color is too dark. In other onboardings, we use color: `var(--studio-gray-60);`.
- The non-selected state of the [SelectCard component](https://www.figma.com/file/eqlepL7MHifK8Bgr1OWqLr/Dotcom-Design-Library?type=design&node-id=1858-1994&t=Slxg49xw3fWg53FN-4) should have a 1px border-width instead of 2px. The description's color should be `color: var(--studio-gray-60);`.
- [The icon](https://www.figma.com/file/eqlepL7MHifK8Bgr1OWqLr/Dotcom-Design-Library?type=design&node-id=47-2902&t=Slxg49xw3fWg53FN-4) in the form-setting-explanation should be `20px`, not `24px`. Additionally, let's increase the paragraph's top margin to `8px` from `6px`.

zaguiini committed:
Do not disable form submissions and display the errors instead

zaguiini committed:
Hide scrollbar

zaguiini commented:
>It could be related to using Calypso live for testing, but the continue button's background color is pink instead of blue.

That's because of the theme used in Calypso. We're using the primary color for the button.

zaguiini committed:
Address coloring issues

zaguiini committed:
Use new endpoint

p-jackson committed:
Import MinimalRequestCartProduct as type only

p-jackson committed:
Reinstate the onChange guard

The refactor in 9124a2f96bee4531f3d97cf1c4e2de0c769aadd4 removed a guard that prevented changes to state until the site had finished loading.

p-jackson reviewed (approved):
Changes are working well. I like the refactor to the options step that allows multiple flows to re-use the nice `options` step name but without having to merge all of the logic into one giant uber step.

I've tested the `videopress` and `site-setup` steps to ensure they're still working as before. No regressions that I could find. There was a change to the behaviour of the `/setup/videopress/options` step in that the "Continue" button is now always enabled rather than waiting for users to enter a valid site title. But this is inline with @javierarce's suggestion above to make the various flows consistent, so I think it's a good change.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7994852

Thank you @zaguiini for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7994852).
